Fixes #26393 - fix uptime calculation for btime

ares opened the pull request:
Thet source of uptime information coming from subscription manager does
not mean uptime but boot time, so we need to convert it in uptime
seconds method.


<!--
Simple description of what is fixed/introduced.

Prerequisites for testing:
* VMware cluster
* Two smart proxies

Tests needed (anticipated impacts):
* Hosts list - mainly power column
* Power status on host page
-->

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [Translating section in the guide]
* Suggest prerequisites for testing and testing scenarios following example above.
(https://projects.theforeman.org/projects/foreman/wiki/Translating)
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* Be nice and respectful

We are running bots that will poke you if you miss an item from the list :-)

-->

ares committed:
Fixes #26393 - fix uptime calculation for btime

Thet source of uptime information coming from subscription manager does
not mean uptime but boot time, so we need to convert it in uptime
seconds method.

ares commented on test/models/host_test.rb:
```diff
@@ -3754,15 +3754,22 @@ def to_managed!
 
     test 'should return uptime and based on backup facts even if there are multiple other facts' do
       host = FactoryBot.create(:host)
-      ansible_uptime_fact = FactoryBot.create(:fact_name, name: 'ansible_uptime_seconds', :type => 'FactName::Ansible')
```
note for the reviewer: this is in fact needed since now we instantiate FactName model, in core we don't have these classes defined

timogoebel commented on app/models/host/base.rb:
```diff
@@ -366,7 +366,12 @@ def render_template(template:, **params)
     def uptime_seconds
-      self.uptime_fact&.value&.to_i
+      fact = self.uptime_fact
+      if fact&.fact_name&.name == 'proc_stat::btime'
```
This is super hacky. Can we please not hardcode subscription manager specific code in core?
Why do we have the fact parser models if we don't use them to parse facts?

Big 👎 from me. Sorry.

timogoebel reviewed (changes requested):
Can we refactor this to put the fact parsing logic into a fact parser?

timogoebel commented:
To explain this a little bit more:

I think we should make the uptime a field on the host model. This then allows us to use scoped search for this.
This field should be updated with every fact import as we do with all other fields during a fact import. Note that we need to clear the field when the host is put in build mode.
Impelemting this feature this way we can add proper tests in the respective plugin's code and not add more technical dept here in core. And the data is still available when the facts expire.

lzap commented:
Let's add new fact called `foreman_uptime` or `foreman::uptime` and modify our importers to set it appropriately.

ekohl commented:
I like the idea of a derived fact and to store it somewhere, but there's some things to consider. For example, what would be the fact source on that fact?

I must also say that I like the boot time from subscription manager a lot more than the uptime from Facter. It doesn't change as often and to get the current uptime you need to calculate anyway. We can normalize uptime into boot time.

lzap commented:
Good point on immutability, let's prefer immutable fact and calculate this in the UI instead!

ares commented:
I agree with Timo, making uptime a first class attribute of host model makes the most sense to me. It's not different to any other information we parse from facts. This will be a bit bigger patch though and this should be fixed in 1.22-stable, so it would be harder and riskier to cherry-pick. Would it be acceptable to take this for 1.22-stable and get the "refactored" PR for develop? Could I base the new work on top of this PR?

I also agree with @ekohl, the new attribute should be boottime instead of uptime to avoid unnecessary updates. But I disagree with derived fact, I belive it should be attribute of host object itself.

timogoebel commented:
> Would it be acceptable to take this for 1.22-stable and get the "refactored" PR for develop?

Fine for me.

> But I disagree with derived fact, I belive it should be attribute of host object itself.

... or a facet.

ares commented:
I was thinking about facet too, but what would that be representing? What other attributes we'd extract? Would it be for facts based values only or generic reported info. Boottime seems as quite generic info, such mac, ip. But the facet may be a way to solve the long standing issue of provisioned vs reported OS.

timogoebel commented:
> I was thinking about facet too, but what would that be representing?

I was thinking about data we store from the hosts that we don't actually need. So data we collect to show it in the UI. Basically, some kind of inventory.

> What other attributes we'd extract?

I'm not sure I would actually extract something but use it to collect more interesting data in the future.

ares committed:
Fixes #26393 - extract uptime to reported facet

A new Reported Data facet is introduced. This should be used for storing
reported information from various fact parsers. Fact Parser now defines
a new optional to implement method, to gather boot time in seconds. Some
fact sources will need to construct this based on uptime.

Hosts can be now search by boot_time search attribute.

ares commented:
I added a second commit that adds a new facet. I'm working on other fact parsers patches. I suppose I'll need to add tests here but I'd appreciate if @timogoebel could confirm this is what you had in mind. Also @ShimShtein this seems to be the first facet in core, any comments from that perspective? Since it's just one attribute, I didn't want to add new tab to host detail page.

Btw with this, people can search using `boot_time > "1 week ago"`. If we're adding more reported data in future (e.g. OS) maybe we should prefix search terms accordingly from the beginning

timogoebel commented on app/models/host/base.rb:
```diff
@@ -1,7 +1,7 @@
   class Base < ApplicationRecord
     KERNEL_RELEASE_FACTS = [ 'kernelrelease', 'ansible_kernel', 'kernel::release' ]
-    UPTIME_FACTS = [ 'system_uptime::seconds', 'ansible_uptime_seconds', 'uptime_seconds', 'proc_stat::btime' ]
+    UPTIME_FACTS = [ 'system_uptime::seconds', 'ansible_uptime_seconds', 'uptime_seconds' ]
```
Is this still used?

ares replied:
These are still uptime facts which can be useful, similar to kernel release fact, now these are really just uptime facts, but if you prefer I can drop.

timogoebel replied:
If we don't use this somewhere, I'd prefer to drop it.

timogoebel commented on app/models/host/base.rb:
```diff
@@ -234,6 +235,14 @@ def set_interfaces(parser)
     end
 
+    def set_reported_data(parser)
+      if parser.boot_time
```
`return unless parser.boot_time`

timogoebel commented on app/models/host/base.rb:
```diff
@@ -234,6 +235,14 @@ def set_interfaces(parser)
 
+    def set_reported_data(parser)
+      if parser.boot_time
+        reported_data_facet = self.reported_data || self.build_reported_data
```
Maybe extract this to a method so you can just call `host.reported_data_facet` and get either the saved object or a new one?
Then we could just do `reported_data_facet.update(boot_time: parser.boot_time)`

timogoebel commented on test/models/host_test.rb:
```diff
@@ -266,6 +266,22 @@ def teardown
+      assert_present host.reported_data.boot_time
+      assert_kind_of DateTime, host.reported_data.boot_time
+
+      sleep 1
```
I hope this is was just for debugging. Note that rails allows you to [travel in time](https://api.rubyonrails.org/v5.2.3/classes/ActiveSupport/Testing/TimeHelpers.html#method-i-travel).

timogoebel reviewed (commented):
@ares: Yeah, that's what I had in mind. Looks good, just some comments.

Note that in the plugin PRs tests are missing.

timogoebel commented:
Note, that the facet should be destroyed when the host enters build mode. Just extend `clear_data_on_build`. A test for this would be great as well.

https://github.com/theforeman/foreman/blob/870c50ad43769b059f363c87d4a2f6c702e5ea9e/app/models/host/managed.rb#L299-L304

ares commented:
updated according to comments, let's see what jenkins says before I squash

timogoebel commented on config/application.rb:
```diff
@@ -273,6 +273,8 @@ class Application < Rails::Application
         child.helper helpers
       end
 
+      Facets.register(HostFacets::ReportedDataFacet, :reported_data)
```
Note, that you need to explicitly set `dependant` to `:destroy`, via `set_dependant :destroy` I believe.

timogoebel commented on db/migrate/20190630034751_create_host_facets_reported_data_facets.rb:
```diff
@@ -0,0 +1,16 @@
+class CreateHostFacetsReportedDataFacets < ActiveRecord::Migration[5.2]
+  def change
+    create_table :host_facets_reported_data_facets do |t|
+      t.integer :host_id
```
`t.reference`?

ezr-ondrej replied:
And you can even
```suggestion
      #index should be default, do not hurt to be explicit though
      t.references :host, type: :integer, foreign_key: true, index: { unique: true } 
```
and you can drop the manual index and foreign key. It looks much cleaner this way.
*Edit*: just read the discussion, see the additional `type:` attribute

timogoebel commented on test/models/host_test.rb:
```diff
@@ -266,6 +266,23 @@ def teardown
       assert host.import_facts(raw['facts'])
     end
+
+    test 'should import facts boot time to reporte data facet' do
```
nit: `to report` (typo)

ares commented:
Thanks @timogoebel, all fixed now

timogoebel commented on db/migrate/20190630034751_create_host_facets_reported_data_facets.rb:
```diff
@@ -1,7 +1,7 @@
   def change
     create_table :host_facets_reported_data_facets do |t|
-      t.integer :host_id
+      t.references :host
```
doesn't this imply the index?

timogoebel replied:
Darn,

```
Column `host_id` on table `host_facets_reported_data_facets` has a type of `bigint(20)`.
This does not match column `id` on `hosts`, which has type `int(11)`.
To resolve this issue, change the type of the `host_id` column on `host_facets_reported_data_facets` to be :integer. (For example `t.integer host_id`).
```

ares replied:
so, reverting this

ezr-ondrej replied:
I would rather see it fixed :( as the foreign_key/primary_key type mismatch is not a good thing to have around.

ezr-ondrej replied:
We can pass `type: :integer` to the references to force it to be an integer.
See https://github.com/rails/rails/blob/a8dfec903bd8cd3b7df9f9d5fbb7100520049302/activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb#L120

ekohl replied:
From experience I can tell that a type mismatch destroys database performance, at least on MySQL. I've seen it ignore indexes and revert to full table scans. :+1: for fixing the root issue.

ares commented:
I reverted the migration tweak and squashed all refs commit, 2 commits now exist (one friendly to cherry-pick, second that fixes it in a better way using facet).

timogoebel reviewed (dismissed):
Code looks good to me.

ares commented:
fixed rubocop, tests should be executed now, @timogoebel thanks for code ack! should I find someone for testing or do you have a katello instance with subscribed host at hand? :-)

timogoebel commented:
> should I find someone for testing or do you have a katello instance with subscribed host at hand? :-)

Some help with testing would be really appreciated.

ezr-ondrej commented:
The uptime is an empty fact for me using `subscription-manager`, @ares please ping me, when you can to help me test it, I got the setup ready :+1:

ares commented:
We can take a look on Monday, but I've seen before, that subscription-manager uptime fact wss not present for some reason. Check /proc/stat if you see boot time info there. If not I can definitely provide a host where I got the info from. It was standard rhel 7 machine IIRC

ezr-ondrej commented on app/models/host/base.rb:
```diff
@@ -234,6 +235,16 @@ def set_interfaces(parser)
+    def set_reported_data(parser)
+      return unless parser.boot_time
+
+      reported_data_facet.boot_time.update!(boot_time: parser.boot_time)
```
I believe you ment:
```suggestion
      reported_data_facet.update!(boot_time: parser.boot_time)
```

ezr-ondrej commented on db/migrate/20190630034751_create_host_facets_reported_data_facets.rb:
```diff
@@ -0,0 +1,10 @@
+class CreateHostFacetsReportedDataFacets < ActiveRecord::Migration[5.2]
+  def change
+    create_table :host_facets_reported_data_facets do |t|
+      t.references :host_id, type: :integer, foreign_key: true, index: { unique: true }
```
:+1: I like we managed to solve it :relaxed:

ares replied:
I had to change `host_id` to `host` though, migrations would fail, I'll push the fix with other updates

ezr-ondrej commented on app/models/host/base.rb:
```diff
@@ -234,6 +233,16 @@ def set_interfaces(parser)
+    def set_reported_data(parser)
+      return unless parser.boot_time
+
+      reported_data_facet.update!(boot_time: parser.boot_time)
```
I would guess you need to do 
```suggestion
      reported_data_facet.update!(boot_time: Time.at(parser.boot_time))
```

ares replied:
yes, this helped, I prefer to store datetime in the DB for the searching, with integer that would be harder...

ezr-ondrej reviewed (changes requested):
In console it already gives the `boot_time` integer from `parser`.
It still doesn't work, though :broken_heart: I would guess it is because Rails do not convert integer to DateTime explicitly, so it filters out the value, and leaves it on original `nil`.
I believe the boot_time fact should give us `Time` already, or if it is giving us timestamp integer, we need eighter:

- do what I suggest inline, or
- make the `boot_time` column to be integer in the database

From the code I feel you expect the value to be integer, so the later would be easier.

ezr-ondrej commented on app/models/host/base.rb:
```diff
@@ -366,7 +375,8 @@ def render_template(template:, **params)
     def uptime_seconds
-      self.uptime_fact&.value&.to_i
+      boot_time = self&.reported_data&.boot_time.to_i
+      boot_time > 0 ? Time.zone.now.to_i - boot_time : nil
```
what about:
```ruby
boot_time = self&.reported_data&.boot_time
Time.zone.now.to_i - boot_time.to_i unless boot_time.nil? # or if boot_time
```

ekohl replied:
How do we know what timezone the boot time is in? I wonder if this can lead to things like -3 hours uptime in the UI.

ares replied:
> what about:

I reverted to `.nil?` and moved `to_i` part lower, I think it's more readable than statement modified at the end of the method

> How do we know what timezone the boot time is in? I wonder if this can lead to things like -3 hours uptime in the UI.

the bootime in /proc/stat is I believe in system time zone, the parsing uses Time.at which assumes boot time in UTC (# of seconds since unix epoch does not have time zone info). When we display it to user, current timezone is used to convert the time in `Time.zone.now`, the same for `boot_time` thanks to active record. When I was testing this, it seemed to work fine.

ezr-ondrej replied:
I believe that the `to_i` removes any information about timezone, what makes maybe the `Time.zone` misleading. `Time.now.to_i` should give the same result.

ekohl commented on app/services/puppet_fact_parser.rb:
```diff
@@ -132,6 +132,10 @@ def support_interfaces_parsing?
   end
 
+  def boot_time
+    Time.zone.now.to_i - facts[:uptime_seconds].to_i
```
The (non-deprecated) Facter 3 fact is `system_uptime`:

```
$ facter --json system_uptime
{
  "system_uptime": {
    "days": 4,
    "hours": 100,
    "seconds": 361191,
    "uptime": "4 days"
  }
}
```

That means you should first look for `facts[:system_uptime][:seconds]` and then fall back. I don't know if the nesting properly converts keys to symbols so probably good to check.

ares replied:
updated

ezr-ondrej commented on app/helpers/hosts_helper.rb:
```diff
@@ -276,7 +276,7 @@ def overview_fields(host)
     fields += [[_("PXE Loader"), host.pxe_loader]] if host.operatingsystem.present? && !host.image_build?
     fields += [[_("Host group"), link_to(host.hostgroup, hosts_path(:search => %{hostgroup_title = "#{host.hostgroup}"}))]] if host.hostgroup.present?
-    fields += [[_("Uptime"), time_ago_in_words(host.uptime_seconds.seconds.from_now)]] if host.uptime_seconds.present?
+    fields += [[_("Boot time"), date_time_relative(host&.reported_data&.boot_time) || _('Not reported')]]
```
Not reported is never going to show up as `date_time_relative` will handle nil value in the react component as 'N/A'.
Could you eighter remove the fallback, or handle it somewhat like (uglier code, but nicer value):
```suggestion
    fields += [[_("Boot time"), (boot_time = host&.reported_data&.boot_time) ? date_time_relative(boot_time) : _('Not reported')]]
```
Or add the option/quark `default_value` to the `date_time_relative` method and propagate it through `mount_date_component` to the react - the best solution :stuck_out_tongue:

ares replied:
I took your suggestion, also squashed all back to just 2 commits since I feel we're getting closer, they should be merged separately at the end

ezr-ondrej commented on app/models/host/base.rb:
```diff
@@ -366,7 +375,8 @@ def render_template(template:, **params)
     def uptime_seconds
-      self.uptime_fact&.value&.to_i
+      boot_time = self&.reported_data&.boot_time
+      boot_time.nil? ? Time.zone.now.to_i - boot_time.to_i : nil
```
This is reversed, test failures are most likely related to that:
```suggestion
      boot_time.present? ? Time.zone.now.to_i - boot_time.to_i : nil
      # OR
      boot_time && Time.zone.now.to_i - boot_time.to_i
```

ares replied:
yeah, I figured while fixing tests :-)

ezr-ondrej reviewed (changes requested):
Looks pretty good already., just two more comments.
I haven't tested `facter` yet, if anyone would like to help out :P

ezr-ondrej reviewed (commented):
Looks good, will retest with puppet after the jenkins ack and I believe we are set :+1:

ares commented:
[test foreman] (broken pipe)

ezr-ondrej commented on test/models/host_test.rb:
```diff
@@ -3749,34 +3766,38 @@ def to_managed!
+      boot_time = 1.day.ago
+      host.create_reported_data(:boot_time => boot_time)
+      refute_nil host.uptime_seconds
+      host.clear_data_on_build
```
I believe, we need to reload after this, the record is destroyed, but still loaded.

ares replied:
in fact it needed more tweaking so that clear_data_on_build would actually work (setting old and build)

ezr-ondrej commented on test/models/host_test.rb:
```diff
@@ -266,6 +266,23 @@ def teardown
+      assert host.import_facts(raw['facts'])
+      first_boot_time = host.reported_data.boot_time
+      refute_nil host.reported_data.boot_time
+      assert_kind_of DateTime, host.reported_data.boot_time
```
This fails, but I don't like asserting exact type match in dynamicaly typed language anyway.
Cant we predict the value and assert the values are equal? The equality should be handled more predictably.
Time is bit tricky, but something like this would maybe do?
```suggestion
      assert (Time.now - host.reported_data.boot_time - raw['facts']['uptime_seconds'].to_i.seconds) < 2.seconds
```

ares replied:
I tried to do something smarter with freeze_time

ezr-ondrej reviewed (commented):
Tests are failing, but it's just because of the tests.

ares commented on test/models/host_test.rb:
```diff
@@ -266,6 +266,25 @@ def teardown
+      freeze_time do
+        host = Host.import_host(raw['name'], 'puppet')
+        assert host.import_facts(raw['facts'])
+        first_boot_time = host.reported_data.boot_time
```
oh dear, fixed cop now

ares commented on test/models/host_test.rb:
```diff
@@ -266,6 +266,26 @@ def teardown
+    test 'should import facts boot time to report data facet' do
+      refute Host.find_by_name('sinn1636.lan')
+      raw = read_json_fixture('facts/facts_with_certname.json')
+      first_boot_time = nil 
```
:see_no_evil:

ezr-ondrej commented on test/models/host_test.rb:
```diff
@@ -266,6 +266,26 @@ def teardown
+      raw = read_json_fixture('facts/facts_with_certname.json')
+      first_boot_time = nil
+      freeze_time do
+        host = Host.import_host(raw['name'], 'puppet')
```
you should define the host variable outside of the block. It is undefined in the other one.

ares replied:
yeah I did the same for first_boot_time already, importing needs to happen while time is frozen... feel free to wait with the review until tests are green, I'm monitoring jenkins output here :-)

ares commented:
I believe that failures are now unrelated and this beast can be merged :-) (please keep 2 separate commits)

ezr-ondrej reviewed (approved):
Code LGTM, works fine.
Thanks @ares :+1: and thanks @timogoebel for most of the review, to @ekohl and @lzap for valuable notes :blush: Merging!
